src/node.cpp

```diff
@@ -1815,77 +1788,30 @@ LocalNode* LocalNode::unserialize(Sync* sync, const string* d)
         type = FILENODE;
     }
 
-    handle fsid = MemAccess::get<handle>(ptr);
-    ptr += sizeof fsid;
-
-    uint32_t parent_dbid = MemAccess::get<uint32_t>(ptr);
-    ptr += sizeof parent_dbid;
-
+    handle fsid;
+    uint32_t parent_dbid;
     handle h = 0;
-    memcpy((char*)&h, ptr, MegaClient::NODEHANDLE);
-    ptr += MegaClient::NODEHANDLE;
-
-    unsigned short localnamelen = MemAccess::get<unsigned short>(ptr);
-    ptr += sizeof localnamelen;
-
-    if (ptr + localnamelen > end)
-    {
-        LOG_err << "LocalNode unserialization failed - name too long";
-        return NULL;
-    }
-
-    const char* localname = ptr;
-    ptr += localnamelen;
+    string localname, shortname;
     uint64_t mtime = 0;
     int32_t crc[4];
     memset(crc, 0, sizeof crc);
+    byte syncable = 1;
+    unsigned char expansionflags[8] = { 0 };
 
-    if (type == FILENODE)
-    {
-        if (ptr + 4 * sizeof(int32_t) > end + 1)
-        {
-            LOG_err << "LocalNode unserialization failed - short fingerprint";
-            return NULL;
-        }
-
-        memcpy(crc, ptr, sizeof crc);
-        ptr += sizeof crc;
-
-        int mtimeSize;
-        if ((mtimeSize = Serialize64::unserialize((byte*)ptr, static_cast<int>(end - ptr), &mtime)) < 0)
-        {
-            LOG_err << "LocalNode unserialization failed - malformed fingerprint mtime";
-            return NULL;
-        }
-        else
-        {
-            ptr += mtimeSize;
-        }
-    }
-
-    char syncable = 1;
-    if (ptr < end)
+    if (!r.unserializehandle(fsid) ||
+        !r.unserializeu32(parent_dbid) || 
+        !r.unserializenodehandle(h) ||
+        !r.unserializestring(localname) ||
+        (type == FILENODE && !r.unserializebinary((byte*)crc, sizeof(crc))) ||
+        (type == FILENODE && !r.unserializecompressed64(mtime)) ||
+        (r.hasdataleft() && !r.unserializebyte(syncable)) ||
+        (r.hasdataleft() && !r.unserializeexpansionflags(expansionflags, 1)) ||
+        (expansionflags[0] && !r.unserializestring(shortname)))
```

Comment:
`shortname` was serialized by using a `serializecstring()`. Shouldn't we use `unserializecstr()` here?

Reply:
Yes good point, I prefer those symmetries too.  Adjusted, thanks